xiilib/_gunicorn/charm_state.py

```diff
@@ -0,0 +1,217 @@
+        self.state_dir = self.base_dir / "state"
+        self.application_log_file = pathlib.Path(f"/var/log/{self.framework}/access.log")
+        self.application_error_log_file = pathlib.Path(f"/var/log/{self.framework}/error.log")
+        self.webserver_config = webserver_config
+        self._wsgi_config = wsgi_config if wsgi_config is not None else {}
+        self._app_config = app_config if app_config is not None else {}
+        self._is_secret_storage_ready = is_secret_storage_ready
+        self._secret_key = secret_key
+        self._database_requirers = database_requirers if database_requirers else {}
+
+    @classmethod
+    def from_charm(  # pylint: disable=too-many-arguments
+        cls,
+        charm: ops.CharmBase,
+        framework: str,
+        wsgi_config: BaseModel,
+        secret_storage: GunicornSecretStorage,
+        database_requirers: dict[str, DatabaseRequires],
+    ) -> "CharmState":
+        """Initialize a new instance of the CharmState class from the associated charm.
+
+        Args:
+            charm: The charm instance associated with this state.
+            framework: The WSGI framework name.
+            wsgi_config: The WSGI framework specific configurations.
+            secret_storage: The secret storage manager associated with the charm.
+            database_requirers: All database requirers object declared by the charm.
+
+        Return:
+            The CharmState instance created by the provided charm.
+        """
+        app_config = {
+            k.replace("-", "_"): v
+            for k, v in charm.config.items()
+            if not any(k.startswith(prefix) for prefix in ("flask-", "webserver-"))
+        }
+        app_config = {k: v for k, v in app_config.items() if k not in wsgi_config.dict().keys()}
+        return cls(
+            framework=framework,
+            wsgi_config=wsgi_config.dict(exclude_unset=True, exclude_none=True),
+            app_config=typing.cast(dict[str, str | int | bool], app_config),
+            database_requirers=database_requirers,
+            webserver_config=WebserverConfig.from_charm(charm),
+            secret_key=(
+                secret_storage.get_secret_key() if secret_storage.is_initialized else None
+            ),
+            is_secret_storage_ready=secret_storage.is_initialized,
+        )
+
+    @property
+    def proxy(self) -> "ProxyConfig":
+        """Get charm proxy information from juju charm environment.
+
+        Returns:
+            charm proxy information in the form of `ProxyConfig`.
+        """
+        http_proxy = os.environ.get("JUJU_CHARM_HTTP_PROXY")
+        https_proxy = os.environ.get("JUJU_CHARM_HTTPS_PROXY")
+        no_proxy = os.environ.get("JUJU_CHARM_NO_PROXY")
+        return ProxyConfig(
+            http_proxy=parse_obj_as(AnyHttpUrl, http_proxy) if http_proxy else None,
+            https_proxy=parse_obj_as(AnyHttpUrl, https_proxy) if https_proxy else None,
+            no_proxy=no_proxy,
+        )
```

Comment:
Can't this be strore as a class attribute to avoid accessing the env_var everytime the proxy property is used ?

Reply:
Yes, it can, but I am not very comfortable to put "runtime" code (accessing environment variables) instead of "compile time" code in metaclass constructor. Maybe it can be cached.

Reply:
Get your point, I try to tend to avoid also login in properties to limit the risks of a property throwing an exceptions. I guess it's fine here as `os.environ` is not a risky thing.
Caching is a good improvement